refactor(directory): Added formatter support for the directory module.

jimleroyer opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!--- To help with semantic versioning the PR title should start with one of the conventional commit types. -->
<!--- The conventional commit types for Semantic PR are: feat, fix, docs, style, refactor, perf, test, build, ci, chore, revert -->

#### Description
<!--- Describe your changes in detail -->

This PR deprecates `symbol` and `style` keys in the directory module to replace it with the `format` key instead.

#### Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
Closes #

Related to issue #1057

#### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#### Screenshots (if appropriate):

![image](https://user-images.githubusercontent.com/805567/79089496-58bfbc80-7d14-11ea-89a9-90417b922494.png)

![image](https://user-images.githubusercontent.com/805567/79089681-fadfa480-7d14-11ea-85ba-577a09a47c61.png)

#### How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, tests ran to see how -->
<!--- your change affects other areas of the code, etc. -->
- [ ] I have tested using **MacOS**
- [ ] I have tested using **Linux**
- [x] I have tested using **Windows**
- [x] I have tested using **Windows Linux Subsystem**

#### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have updated the documentation accordingly.
- [x] I have updated the tests accordingly.

jimleroyer commented on src/configs/directory.rs:
```diff
@@ -21,8 +19,7 @@ impl<'a> RootModuleConfig<'a> for DirectoryConfig<'a> {
             use_logical_path: true,
-            prefix: "in ",
-            style: Color::Cyan.bold(),
+            format: "[$path](cyan bold) ",
```
Shall we add the `in ` prefix in the `format` string? I didn't see that prefix locally and my personal configuration didn't overwrite it. Hence to have similar results, I didn't add it back here.

matchai replied:
Ooh... You've landed on an interested edge-case we didn't account for. The first prefix on each line isn't shown.

Yeah, I think it's best if we remove it with format strings.

jimleroyer replied:
Ah that would explain why I didn't see it locally!

What will happen to prefix/suffix support? Will this be removed once the formatter is fully implemented?

matchai replied:
Yep, that's the plan.

jimleroyer commented on src/modules/directory.rs:
```diff
@@ -64,33 +63,33 @@ pub fn module<'a>(context: &'a Context) -> Option<Module<'a>> {
-            style: None,
-        },
-    );
+    let final_dir_string = format!("{}{}", fish_prefix, truncated_dir_string);
```
I did test with the `fish_style_pwd_dir_length` configuration value > 0, and that seemed to work when comparing to the official build.

jimleroyer commented:
The tests around prefix are currently failing. I updated these tests locally and it's working now, but because I added some comments around support/prefix support planning, I just want to know before a push what we plan to do with these.

heyrict commented on src/configs/directory.rs:
```diff
@@ -9,8 +8,7 @@ pub struct DirectoryConfig<'a> {
     pub use_logical_path: bool,
-    pub prefix: &'a str,
-    pub style: Style,
+    pub format: &'a str,
```
Same as #1066

jimleroyer replied:
I reinstated the `style` configuration.

heyrict commented on src/modules/directory.rs:
```diff
@@ -64,33 +63,33 @@ pub fn module<'a>(context: &'a Context) -> Option<Module<'a>> {
-    );
+    let final_dir_string = format!("{}{}", fish_prefix, truncated_dir_string);
+
+    let formatter = if let Ok(formatter) = StringFormatter::new(config.format) {
+        formatter.map(|variable| match variable {
+            "path" => Some(final_dir_string.clone()),
+            _ => None,
+        })
+    } else {
+        log::warn!("Error parsing format string in `directory.format`");
+        return None;
+    };
```
Ditto

jimleroyer replied:
And that was updated with latest code changes.

heyrict commented on tests/testsuite/directory.rs:
```diff
@@ -90,11 +90,12 @@ fn root_directory() -> io::Result<()> {
     Ok(())
 }
 
+// TODO: Remove this test once formatter support is fully implemented and prefix support has been removed.
```
I think it is the time :)

jimleroyer replied:
Alrighty, just removed it. 👍

heyrict commented on docs/config/README.md:
```diff
@@ -446,13 +446,12 @@ it would have been `nixpkgs/pkgs`.
-| `prefix`            | `"in "`       | Prefix to display immediately before the directory.                              |
-| `style`             | `"bold cyan"` | The style for the module.                                                        |
-| `disabled`          | `false`       | Disables the `directory` module.                                                 |
+| Variable                    | Default                    | Description                                                                      |
```
Variable should be changed to Option to avoid naming conflicts with format string variables.

jimleroyer replied:
Changed..

heyrict commented on docs/config/README.md:
```diff
@@ -446,13 +446,12 @@ it would have been `nixpkgs/pkgs`.
+| --------------------------- | -------------------------- | -------------------------------------------------------------------------------- |
+| `truncation_length`         | `3`                        | The number of parent folders that the current directory should be truncated to.  |
+| `truncate_to_repo`          | `true`                     | Whether or not to truncate to the root of the git repo that you're currently in. |
+| `format`                    | `"[$path]($style) "` | The format for the module.                                                       |
```
It seems that you missed a style option  here.

jimleroyer replied:
Added it..

heyrict commented on src/modules/directory.rs:
```diff
@@ -70,33 +69,44 @@ pub fn module<'a>(context: &'a Context) -> Option<Module<'a>> {
+            .map(|variable| match variable {
+                // This may result in multiple calls to `get_module_version` when a user have
+                // multiple `$version` variables defined in `format`.
+                "path" => Some(Ok(final_dir_string.clone())),
```
I think returning a reference should work now.

By the way the comments are of non sense :)

jimleroyer replied:
I removed the comments, sorry, damn copy/paste. 😬 

Sending a reference works. 👍  (I'll update other PRs as well with this pattern.)

heyrict commented on docs/config/README.md:
```diff
@@ -447,13 +447,13 @@ it would have been `nixpkgs/pkgs`.
+| `truncation_length` | `3`                  | The number of parent folders that the current directory should be truncated to.  |
+| `truncate_to_repo`  | `true`               | Whether or not to truncate to the root of the git repo that you're currently in. |
+| `format`            | `"[$path]($style) "` | The format for the module.                                                       |
+| `style`             | `"bold red"`         | The style for the module.                                                        |
```
```suggestion
| `style`             | `"bold cyan"`        | The style for the module.                                                        |
```

heyrict commented on docs/config/README.md:
```diff
@@ -462,6 +462,7 @@ it would have been `nixpkgs/pkgs`.
 | --------------------------- | ------- | ---------------------------------------------------------------------------------------- |
 | `fish_style_pwd_dir_length` | `0`     | The number of characters to use when applying fish shell pwd path logic.                 |
 | `use_logical_path`          | `true`  | Displays the logical path provided by the shell (`PWD`) instead of the path from the OS. |
+| style                       |         | Mirrors the value of option `style`                                                      |
```
I think this should be added to the variable section.

jimleroyer replied:
Not sure I understand. This is the variable section and you refer to `style`?

heyrict replied:
No, this is a table for advanced configurations. You may want to write another table for the details of format-string variables.

jimleroyer replied:
Oh I see. I created a new section and moved `style` in there (along with `path` as well, part of the variables for the string format).

I also renamed that advanced option column header to `Advanced Option` instead of `Variable`.

jimleroyer commented:
@heyrict It seems this PR still have changes requested but I see that all comments have been resolved. Can you point to me which changes would be remaining, if any? Thank you!
